src/lib/consensus/proof_of_stake.ml

```diff
@@ -134,21 +132,28 @@ module Data = struct
   module Epoch_and_slot = struct
     type t = Epoch.t * Slot.t [@@deriving eq, sexp]
 
-    let of_time_exn tm : t =
-      let epoch = Epoch.of_time_exn tm in
-      let time_since_epoch = Time.diff tm (Epoch.start_time epoch) in
+    let of_time_exn tm ~genesis_state_timestamp ~epoch_duration
```

Comment:
I'm a little concerned with the increased complexity in this file w.r.t. the addition of parameters to most of the functions here.

I wonder if it could be nicer to stage the `Constants.t` value via some first-class module instantiation or adding a context parameter to all the functions that depend on constants .

This would be a large change to make at this point, so I'd like for at least one other person to second this before going forward with that change. An alternative is to address it in a follow up too.

Reply:
What do you mean by -- stage the Constants.t value?

Reply:
I think brandon is suggesting either making everything a functor over `sig val constants : Constants.t end` so you don't have to have additional parameters on all the functions or just having one additional parameter `constants:Constants.t` on each function rather than passing them individually.

Reply:
I decided against functors because that might make module initialization complex again since we recently defuntored most of them. But maybe it won't be that bad with just one module that doesn't depend on any other modules?
I think all of the exposed functions (in `Consensus.Intf`) have `constants:Constants.t`. Possible that I might have missed a few. But within the consensus library I did not follow that rule.